### ISSUE TITLE
[Question] Cluster support planned?

### ISSUE THREAD

**Issue by asciidisco** [mafintosh/turbo-http#11]
Hi there.
First of all, turbo-net & turbo-http are amazing, thank you very much :)

If I´m not mistaken, turbo-http doesn't work in cluster mode yet, as it doesn't implement the same "behind the scenes magick" as the default http module does to "share" an open port.

I´ve just, very naively, tried to map the example from the docs 1:1 - https://nodejs.org/docs/latest/api/cluster.html

So, chances are high that I´m just doing something wrong here, but if not, I´d just simply ask if it is planned to implement cluster compat., or if I should go down the route of forking instances & using a traditional load balancer instead of cluster mode?

**Comment by jacktuck** [comment]
> If I´m not mistaken, turbo-http doesn't work in cluster mode yet, as it doesn't implement the same "behind the scenes magick" 

I _think_ you're right here, just ran into this myself too. So +1

**Comment by mafintosh** [comment]
Yes turbo-net does not do any clustering. I'm not sure what's involved to get that working.

**Comment by asciidisco** [comment]
@mafintosh Thank you for the answer, it´s not that critical, as I don't want to use it for a production product yet & I´m just toying around with it.

I´ll see if I can find some clues in the node sources, but I doubt that this is a trivial implementation.

**Comment by jacktuck** [comment]
haha @asciidisco i've been doing exactly that and... it's been quite the rabbit hole 

e.g. https://github.com/nodejs/node/blob/212de3c5ec429a580d2e79ce3c2516b93b52b8f5/lib/_http_common.js#L264 😱

**Comment by asciidisco** [comment]
@jacktuck https://github.com/nodejs/node/blob/master/lib/internal/cluster/round_robin_handle.js

**Comment by asciidisco** [comment]
The way it is implemented in node is quite understandable. It looks like one could create a separate package that could just wrap `turbo-http` and make it cluster-able.

If I´m not missing any important point, I´d say that if it can be done as a separate package, it should be done this way, as every implementation in `turbo-http` would have a performance impact of some sort, even if users wouldn't need the functionality.

Therefore I´m fine with closing this. If >1 instances are needed, people could still use the "classic" load balancer approach.

**Comment by jacktuck** [comment]
> The way it is implemented in node is quite understandable

Can you give me synopsis of how node's doing it? I think the bit you linked to is related to distributing workers onto 'random' ports

Yeah i do wonder what the performance hit would be

**Comment by asciidisco** [comment]
@jacktuck It´s explained in the docs itself (better than I could ever do it): https://nodejs.org/api/cluster.html#cluster_how_it_works

**Comment by aichholzer** [comment]
If you need clustering, you can use [@rayo/storm](https://github.com/GetRayo/rayo.js/tree/master/packages/storm) which will basically spawn your worker (server) function over all available CPU cores (you can also limit the number).

Something along the lines of:
```js
const rayo = require('rayo');
const turbo = require('turbo-http');
const server = turbo.createServer();

rayo({ port: 5050, server, storm: {} })
  .get('/', (req, res) => res.end('Thunderstruck by rayo, turbo and storm..!'))
  .start((address) => {
    console.log(`Up on port ${address.port}`);
  });
```

In this case `rayo.js` will, essentially, just provide the routing capabilities. Also, in this example, you don't need to require `@rayo/storm` since `rayo` already comes bundled with it, you just need to turn it on.

**Comment by jfrconley** [comment]
I submitted a pull request to turbo-net that enables cluster support.
https://github.com/mafintosh/turbo-net/pull/16

**Comment by JCWolf** [comment]
Hi. First of all, great work here. 
Are there any updates on this?

**Comment by dalisoft** [comment]
Any news?
